File: 142-linked-list-cycle-ii/linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast != NULL && fast->next != NULL)
        {
            slow = slow->next;
            fast = fast->next->next;
            if(slow == fast){
                int count = 0;
                slow = head;
                while(slow!=fast){
                    slow=slow->next;
                    fast=fast->next;
                    count++;
                }
                return slow;
            }
        }
        return NULL;
    }
};

File: 142-linked-list-cycle-ii/linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        int length = 0;
        ListNode fast = head;
        ListNode slow = head;
        while(fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast){
                // finding length of cycle
                do{
                    slow = slow.next;
                    length++;
                } while (slow != fast);
                break;
            }
        }
        
        if(length == 0)
        {
            return null;
        }

        ListNode first = head;
        ListNode second = head;
        while(length>0)
        {
            first = first.next;
            length--;
        }

        // To find where the cycle meet:
        while(first != second)
        {
            first = first.next;
            second = second.next;
        }
        return second;
    }
}